src/nbl/asset/utils/IShaderCompiler.cpp

```diff
@@ -0,0 +1,384 @@
+{}
+
+std::string IShaderCompiler::CFileSystemIncludeLoader::getInclude(const system::path& searchPath, const std::string& includeName) const
+{
+    system::path path = searchPath / includeName;
+    if (std::filesystem::exists(path))
+        path = std::filesystem::canonical(path);
+
+    core::smart_refctd_ptr<system::IFile> f;
+    {
+        system::ISystem::future_t<core::smart_refctd_ptr<system::IFile>> future;
+        m_system->createFile(future, path.c_str(), system::IFile::ECF_READ);
+        f = future.get();
+        if (!f)
+            return {};
+    }
+    const size_t size = f->getSize();
+
+    std::string contents(size, '\0');
+    system::IFile::success_t succ;
+    f->read(succ, contents.data(), 0, size);
+    const bool success = bool(succ);
+    assert(success);
+
+    return contents;
+}
+
+IShaderCompiler::CIncludeFinder::CIncludeFinder(core::smart_refctd_ptr<system::ISystem>&& system) 
+    : m_defaultFileSystemLoader(core::make_smart_refctd_ptr<CFileSystemIncludeLoader>(std::move(system)))
+{
+    addSearchPath("", m_defaultFileSystemLoader);
+}
+
+// ! includes within <>
+// @param requestingSourceDir: the directory where the incude was requested
+// @param includeName: the string within <> of the include preprocessing directive
+std::string IShaderCompiler::CIncludeFinder::getIncludeStandard(const system::path& requestingSourceDir, const std::string& includeName) const
+{
+    std::string ret = tryIncludeGenerators(includeName);
+    if (ret.empty())
+        ret = trySearchPaths(includeName);
+    if (ret.empty())
+        ret = m_defaultFileSystemLoader->getInclude(requestingSourceDir.string(), includeName);
+    return ret;
+}
+
+// ! includes within ""
+// @param requestingSourceDir: the directory where the incude was requested
+// @param includeName: the string within "" of the include preprocessing directive
+std::string IShaderCompiler::CIncludeFinder::getIncludeRelative(const system::path& requestingSourceDir, const std::string& includeName) const
+{
+    std::string ret = m_defaultFileSystemLoader->getInclude(requestingSourceDir.string(), includeName);
+    if (ret.empty())
+        ret = trySearchPaths(includeName);
+    return ret;
+}
+
+void IShaderCompiler::CIncludeFinder::addSearchPath(const std::string& searchPath, const core::smart_refctd_ptr<IIncludeLoader>& loader)
+{
+    if (!loader)
+        return;
+    m_loaders.push_back(LoaderSearchPath{ loader, searchPath });
+}
+
+void IShaderCompiler::CIncludeFinder::addGenerator(const core::smart_refctd_ptr<IIncludeGenerator>& generator)
+{
+    if (!generator)
+        return;
+
+    auto itr = m_generators.begin();
+    for (; itr != m_generators.end(); ++itr)
+    {
+        auto str = (*itr)->getPrefix();
+        if (str.compare(generator->getPrefix()) <= 0) // Reverse Lexicographic Order
+            break;
+    }
+    m_generators.insert(itr, generator);
```

Comment:
`lower_bound` takes a comparator functor argument btw, you could make this binary-insertion sorted

Reply:
ok, good idea, I'll use lower_bound instead of my custom for